### PR TITLE
feat(instantsearch.js): introduce `carousel` template

### PR DESCRIPTION
**Pull request by sarahdayan**
This introduces the `carousel` template and implements the `layout` template in Recommend widgets.

https://algolia.atlassian.net/browse/FX-2818

### CODE REVIEW

File: packages/instantsearch.js/src/templates/__tests__/carousel.test.tsx [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+            title="Next"
+          >
+            <p>
+              Next
+            </p>
+          </button>
+        </div>
+      </div>
+    `);
+  });
```
Comment (Haroenv) [inline packages/instantsearch.js/src/templates/__tests__/carousel.test.tsx:226]: I don't think we need to add that or think about it yet, but when hit is supported, this needs insights tests and functionality too.

File: packages/instantsearch.js/src/templates/carousel/carousel.tsx [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  'items'
+> & {
+  templates: {
+    item?: CarouselUiProps<TObject>['itemComponent'];
+  };
+};
+
+export function carousel<TObject extends Record<string, unknown>>({
+  cssClasses,
+  templates,
```
Comment (Haroenv) [inline packages/instantsearch.js/src/templates/carousel/carousel.tsx:61]:
```suggestion
  templates = {},
```

if code ~ like this, you don't need to check `templates?.` later

File: packages/instantsearch.js/src/widgets/frequently-bought-together/frequently-bought-together.tsx [hunk 3]
```diff
@@ -111,6 +111,33 @@ const renderer =
...
+      templates.layout
+        ? (data) => (
+            <TemplateComponent
+              {...renderState.templateProps}
+              templateKey="layout"
+              rootTagName="fragment"
+              data={{
+                items: data.items,
+                templates: {
+                  item: templates.item
```
Comment (Haroenv) [inline packages/instantsearch.js/src/widgets/frequently-bought-together/frequently-bought-together.tsx:124]: surely there's always at least a default item template?

**Review by Haroenv — COMMENTED** [review]
🚀

File: packages/instantsearch.js/src/templates/__tests__/carousel.test.tsx (hunk 1 above)
Comment (sarahdayan) [reply packages/instantsearch.js/src/templates/__tests__/carousel.test.tsx:226]: Yes, we can accept `sendEvent` in there when it's time.

File: packages/instantsearch.js/src/widgets/frequently-bought-together/frequently-bought-together.tsx (hunk 3 above)
Comment (sarahdayan) [reply packages/instantsearch.js/src/widgets/frequently-bought-together/frequently-bought-together.tsx:124]: The default item template is defined in the UI component. At the widget level, the `itemComponent` can be undefined, which will be forwarded as-is to the UI component and be defaulted there.

File: packages/instantsearch.js/src/templates/carousel/carousel.tsx (hunk 2 above)
Comment (sarahdayan) [reply packages/instantsearch.js/src/templates/carousel/carousel.tsx]: Yup! Done in [77f0549](https://github.com/algolia/instantsearch/pull/6314/commits/77f0549d51380c2fc7e11659806810a107bd88e0).

**Review by dhayab — APPROVED** [review]
Awesome! We can probably update `trendingItems()` the js getting-started example to use the carousel layout. We can also wait until we release instantsearch.css (we load themes from a CDN in the examples).

**Review by Haroenv — APPROVED** [review]

**Pull request merged**